src/core/RenderManager.cpp

```diff
@@ -0,0 +1,214 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include <QFileInfo>
+#include <QDebug>
+#include <QDir>
+
+#include "RenderManager.h"
+#include "Song.h"
+#include "BBTrackContainer.h"
+#include "BBTrack.h"
+#include "debug.h"
+
+RenderManager::RenderManager(
+		const Mixer::qualitySettings & qualitySettings,
+		const ProjectRenderer::OutputSettings & outputSettings,
+		ProjectRenderer::ExportFileFormats fmt,
+		QString outputPath) :
+	m_qualitySettings(qualitySettings),
+	m_outputSettings(outputSettings),
+	m_format(fmt),
+	m_outputPath(outputPath),
+	m_activeRenderer(NULL)
+{
+}
+
+RenderManager::~RenderManager()
+{
+	delete m_activeRenderer;
+}
+
+void RenderManager::abortProcessing()
+{
+	if ( m_activeRenderer ) {
+		m_activeRenderer->abortProcessing();
+	}
+	restoreMutedState();
+}
+
+// Called to render each new track when rendering tracks individually.
+void RenderManager::renderNextTrack()
+{
+	delete m_activeRenderer;
+	m_activeRenderer = NULL;
+
+	if( m_tracksToRender.isEmpty() )
+	{
+		// nothing left to render
+		restoreMutedState();
+		emit finished();
+	}
+	else
+	{
+		// pop the next track from our rendering queue
+		Track* renderTrack = m_tracksToRender.back();
+		m_tracksToRender.pop_back();
+
+		// mute everything but the track we are about to render
+		for( auto it = m_unmuted.begin(); it != m_unmuted.end(); ++it )
+		{
+			(*it)->setMuted( (*it) != renderTrack );
+		}
+
+		// for multi-render, prefix each output file with a different number
+		int trackNum = m_tracksToRender.size() + 1;
+
+		// create a renderer for this track
+		m_activeRenderer = new ProjectRenderer(
+				m_qualitySettings,
+				m_outputSettings,
+				m_format,
+				pathForTrack(renderTrack, trackNum));
+
+		if ( m_activeRenderer->isReady() )
+		{
+			// pass progress signals through
+			connect( m_activeRenderer, SIGNAL( progressChanged( int ) ),
+					this, SIGNAL( progressChanged( int ) ) );
+
+			// when it is finished, render the next track
+			connect( m_activeRenderer, SIGNAL( finished() ),
+					this, SLOT( renderNextTrack() ) );
+
+			m_activeRenderer->startProcessing();
+		}
+		else
+		{
+			qDebug( "Renderer failed to acquire a file device!" );
+			renderNextTrack();
+		}
+	}
+}
+
+// Render the song into individual tracks
+void RenderManager::renderTracks()
+{
+	const TrackContainer::TrackList & tl = Engine::getSong()->tracks();
+
+	// find all currently unnmuted tracks -- we want to render these.
+	for( auto it = tl.begin(); it != tl.end(); ++it )
+	{
+		Track* tk = (*it);
```

Comment:
In the future, feel free to use range-based iterators, i.e. `for (Track * tk : tl) { ... }`. Saves a bit of redundancy.